server/blob/blob-deduplicating/src/test/scala/GCPropertiesTest.scala

```diff
@@ -0,0 +1,283 @@
+import org.apache.james.blob.api.{BlobId, TestBlobId}
+import org.scalacheck.Prop.forAll
+import org.scalacheck.{Arbitrary, Gen, Properties}
+
+case class Generation(id: Long) {
+  def previous: Generation = copy(id - 1)
+  def next: Generation = copy(id + 1)
+
+  def collectibles(targetedGeneration: Generation): Set[Generation] =
+    ((id + 1) to (targetedGeneration.id - 2)).toSet.map(Generation(_))
+}
+
+object Generation {
+  implicit val ord: Ordering[Generation] = Ordering.by(unapply)
+}
+
+case class Iteration(id: Long, processedGenerations: Set[Generation]) {
+  def next(generations: Set[Generation]): Iteration = Iteration(id + 1, generations)
+  def lastGeneration: Generation = processedGenerations.max
+}
+
+object Iteration {
+  def initial: Iteration = Iteration(0, Set(Generation(-1)))
+}
+
+case class ExternalID(id: String) // TODO
+
+sealed trait Event {
+  def blob: BlobId
+  def externalId: ExternalID
+  def generation: Generation
+}
+
+case class Reference(externalId: ExternalID, blobId: BlobId, generation: Generation) extends Event {
+  override def blob: BlobId = blobId
+}
+
+case class Deletion(generation: Generation, reference: Reference) extends Event {
+  override def blob: BlobId = reference.blob
+  override def externalId: ExternalID = reference.externalId
+}
+
+case class Report(iteration: Iteration,
+                  blobsToDelete: Set[(Generation, BlobId)]
+                 )
+
+case class State(blobs: Set[BlobId],
+                 references: Map[Generation, Set[Reference]],
+                 deletions: Map[Generation, Set[Deletion]]) {
+  private lazy val referencedBlobsAcrossGenerations: Map[Generation, ReferencedBlobs] = {
+    val maxGeneration = (references.keys ++ deletions.keys).max
+    val minGeneration = (references.keys ++ deletions.keys).min
+    val initialRefs = ((-1L) until minGeneration.id).map(Generation(_)).map((_, ReferencedBlobs(Map()))).toMap
+    (minGeneration.id to maxGeneration.id)
```

Comment:
I had hard time reading this code, it deserves method more extraction/naming care